Python/Product/Analysis/LanguageServer/Server.Completion.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.PythonTools.Interpreter;
+using Microsoft.PythonTools.Parsing;
+
+namespace Microsoft.PythonTools.Analysis.LanguageServer {
+    public sealed partial class Server {
+        private TaskCompletionSource<bool> _analysisCts;
+
+        public override async Task<CompletionList> Completion(CompletionParams @params) {
+            await _analyzerCreationTask;
+            await IfTestWaitForAnalysisCompleteAsync();
+
+            var uri = @params.textDocument.uri;
+            // Make sure document is enqueued for processing
+            var openFile = _openFiles.GetDocument(uri);
+
+            _projectFiles.GetAnalysis(@params.textDocument, @params.position, @params._version, out var entry, out var tree);
+            TraceMessage($"Completions in {uri} at {@params.position}");
+
+            tree = GetParseTree(entry, uri, CancellationToken, out _) ?? tree;
+            // Give analysis opportunity to complete
+            if (!entry.IsComplete) {
+                entry.OnNewAnalysis += OnNewAnalysis;
+                _analysisCts = new TaskCompletionSource<bool>();
+                _analysisCts.Task.Wait(300);
+            }
+
+            var analysis = entry?.Analysis;
+            if (analysis == null) {
+                TraceMessage($"No analysis found for {uri}");
+                return new CompletionList();
+            }
+
+            var opts = GetOptions(@params.context);
+            var ctxt = new CompletionAnalysis(analysis, tree, @params.position, opts, this);
+            var members = ctxt.GetCompletionsFromString(@params._expr) ?? ctxt.GetCompletions();
+
+            if (_settings.SuppressAdvancedMembers) {
+                members = members.Where(m => !m.label.StartsWith("__"));
```

Comment:
This check should also include `EndsWithOrdinal("__")`. Member prefixed with `__` are protected against accidental overwriting by subclasses (via automatic name mangling), but they are not "advanced" members.

Reply:
OK